guppylang/tys/errors.py

```diff
@@ -0,0 +1,127 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, ClassVar
+
+from guppylang.diagnostic import Error, Help, Note
+
+if TYPE_CHECKING:
+    from guppylang.definition.parameter import ParamDef
+    from guppylang.tys.ty import Type
+
+
+@dataclass(frozen=True)
+class WrongNumberOfTypeArgsError(Error):
+    title: ClassVar[str] = ""  # Custom implementation in `rendered_title`
+    expected: int
+    actual: int
+    type_name: str
+
+    @property
+    def rendered_title(self) -> str:
+        if self.expected > self.actual:
+            return "Missing type arguments"
+        elif 0 == self.expected < self.actual:
```

Comment:
nit: this case is a bit confusing - I mean, the assumption is that `expected>=0 and actual>=0 and expected!=actual` in which case the `< self.actual` here is redundant. I'd just be consistent with `rendered_span_label` i.e. check `self.expected == 0` *first* and then the other two.